gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+                         0.0  /* default value */,
+                         G_PARAM_READWRITE);
+
+  _signals[SIG_FINISHED] =
+    g_signal_newv ("finished",
+                   G_TYPE_FROM_CLASS (gobject_class),
+                   G_SIGNAL_RUN_LAST | G_SIGNAL_NO_RECURSE | G_SIGNAL_NO_HOOKS,
+                   NULL /* closure */,
+                   NULL /* accumulator */,
+                   NULL /* accumulator data */,
+                   NULL /* C marshaller */,
+                   G_TYPE_NONE /* return_type */,
+                   0     /* n_params */,
+                   NULL  /* param_types */);
+
+  g_object_class_install_properties (gobject_class,
+                                     N_PROPERTIES,
+                                     _props);
+}
+
+EosReformatter *
+eos_reformatter_new (const gchar *image, const gchar *signature, const gchar *device)
+{
+  return EOS_REFORMATTER (g_object_new (EOS_TYPE_REFORMATTER,
+                                      "image", image,
+                                      "device", device,
+                                      "signature", signature,
+                                      NULL));
+}
+
+/* Private API */
+
+static void
+eos_reformatter_maybe_finish (EosReformatter *reformatter)
+{
+  if (reformatter->finished)
+    return;
+
+  if (reformatter->error == NULL && reformatter->gpg > 0)
+    return;
+
+  if (reformatter->error == NULL && reformatter->progress < 1.0)
+    return;
+
+  reformatter->finished = TRUE;
+
+  g_signal_emit (reformatter, _signals[SIG_FINISHED], 0);
+}
+
+static EosAlignedBuffer *
+eos_reformatter_get_free_buffer (EosReformatter *reformatter, gboolean exhaust)
+{
+  /* If the exhaust flag is not set, we only allow half of the buffers to be
+   * reserved. This is mainly to avoid reading queue to hog free buffers.
+   */
+  int half_buffers = reformatter->total_buffers / 2;
+  EosAlignedBuffer *buf = NULL;
+
+  if (exhaust || g_async_queue_length (reformatter->free_queue) < half_buffers)
+    {
+      buf = g_async_queue_try_pop (reformatter->free_queue);
+    }
+
+  /* A free buffer was available, return it */
+  if (buf != NULL)
+    {
+      buf->len = 0;
+      return buf;
+    }
+
+  /* Initialize our pool, if it's not created yet */
+  if (reformatter->pool == NULL)
+    {
+      reformatter->pool = g_new0(guchar, reformatter->total_buffers * (reformatter->buffer_size + reformatter->page_size));
+    }
+
+  /* A bit awkwardly, exhaust doesn't apply when we create new buffers */
+  if (reformatter->used_buffers < reformatter->total_buffers)
+    {
+      /* No free buffers, but we can create one */
+      gsize offset = reformatter->used_buffers * (reformatter->buffer_size + reformatter->page_size);
+      guchar *unaligned = reformatter->pool + offset;
+      gsize page_size = reformatter->page_size;
+
+      buf = g_new0(EosAlignedBuffer, 1);
+      buf->ptr = (guchar*) (((gintptr) (unaligned + page_size)) & (~(page_size - 1)));
```

Comment:
We still need to carry the size information though, since the decompression step produces variable-sized output (unless we make that step complex by always collecting fixed size buffers until passing it on). We could allocate the memory for the buffers with posix_memalign, though I'm not sure if that's better or worse than a large contiguous blob of memory.